TicketUtil.py

```diff
@@ -119,6 +118,123 @@ def close_requests_session(self):
+        self.auth_url = '{0}/rest/login'.format(self.base_url)
+
+        # Call our parent class's init method which creates our requests session.
+        super(BugzillaTicket, self).__init__(project_key, ticket_id)
+
+
+    def create_ticket_parameters(self, options_dict):
+
+        params = {"product": str(self.project_key)}
+
+        # Iterate through our options and add them to the params dict.
+        for key, value in options_dict.items():
+            params.update({str(key): str(value)})
+        return params
+
+    def create_requests_session(self):
+        if self.auth != 'rest':
+            return super()
+        else:
+            try:
+                s = requests.Session()
+                r = s.get(self.auth_url, params=self.credentials, verify=False)
+                r.raise_for_status()
+                self.checkForBZError(r)
+                resp = r.json()
+                self.token = resp['token']
+                logging.debug("Create requests session: Status Code: {0}".format(r.status_code))
+                logging.info("Successfully authenticated to {0} with token: {1}".format(self.ticketing_tool, self.token))
+
+                return s
+            # We log an error if authentication was not successful, because rest of the HTTP requests will not succeed.
+
+            except requests.RequestException as e:
+                logging.error("Error authenticating to {0}. No valid credentials were provided.".format(self.auth_url))
+                logging.error(e.args[0])
+
+
+    def checkForBZError(self, bzRessponse):
+        if 'error' in bzRessponse.json() and bzRessponse.json()['error']:
+            raise Exception(bzRessponse.json()['message'] if 'message' in bzRessponse.json() else 'Unknown API error')
+
+
+    def create_ticket(self, params):
+        """
+        Tries to create the ticket through the ticketing tool's API.
+        Retrieves the ticket_id and creates the ticket_url.
+        :param params: The payload to send in the POST request.
+        :return:
+        """
+        # Attempt to create ticket.
+        try:
+            r = self.s.post(self.rest_url, params={'token': self.token}, json=params)
+
+            r.raise_for_status()
+            logging.debug("Create ticket: Status Code: {0}".format(r.status_code))
+
+            self.checkForBZError(r)
+
+            self.ticket_id = r.json()['id']
+            self.ticket_url = self.craft_ticket_url()
+            logging.info("Created ticket {0} - {1}".format(self.ticket_id, self.ticket_url))
+
+        # If ticket creation is not successful, log an error.
+        except requests.RequestException as e:
+            logging.error("Error creating ticket")
+            logging.error(e.args[0])
+
+    def craft_ticket_url(self):
+        ticket_url = None
+
+        if self.ticket_id:
+            ticket_url = "{0}/show_bug.cgi?id={1}".format(self.base_url, self.ticket_id)
+            print(ticket_url)
```

Comment:
done :) sorry somehow the old code was pushed as I was learning git, your suggestions are considered Tom.